### DISCUSSION TITLE
How to parse query string or query params of a Request object?

### DISCUSSION THREAD

**Discussion by postmodern** [socketry/async-http#164] [Q&A] [answered]
I just noticed that the query portion (ex: `?q=1`) of the request URI is not separated from the request path (ex: `/foo`). Instead `Request#path` includes both the path and query string (ex: `/foo?q=1`). Is this intentional? Is there some kind of method I didn't notice which parses the request URI into path and query parts?

**Comment by ioquatix** [comment, marked as answer]
I did not want to mix URL specifications with HTTP specifications, however there is some support for it in the `protocol-http` gem. Probably we should make a more extensive gem for this - e.g. parsing out form data and so on.

https://socketry.github.io/protocol-http/guides/getting-started/index#url-parsing

I know this isn't probably the answer you were looking for, but URL parsing (and caching, and form data) is a Pandora's box I didn't want to open in the core interfaces for request/response, at least not until I had a good idea of how to do it *correctly*.

**Reply by postmodern** [reply]
For my [use case](https://github.com/ronin-rb/ronin-listener-http), I can parse the request URI path using `Protocol::HTTP::Reference` and `Protocol::HTTP::URL` and initialize my own `Request` object. It would be nice if async-http had an API similar to [Rack::Request](https://rubydoc.info/gems/rack/Rack/Request).
